plugins/inventory/ec2.py

```diff
@@ -379,23 +383,33 @@ def get_rds_instances_by_region(self, region):
                 for instance in instances:
                     self.add_rds_instance(instance, region)
         except boto.exception.BotoServerError, e:
+            if e.error_code == 'AuthFailure':
+                self.display_auth_error()
+
             if not e.reason == "Forbidden":
                 print "Looks like AWS RDS is down: "
                 print e
                 sys.exit(1)
 
-    def get_instance(self, region, instance_id):
-        ''' Gets details about a specific instance '''
-        if self.eucalyptus:
-            conn = boto.connect_euca(self.eucalyptus_host)
-            conn.APIVersion = '2010-08-31'
+    def display_auth_error(self):
+        ''' Raise an error with an informative message if there is an issue authenticating'''
+        errors = ["Authentication error retrieving ec2 inventory."]
+        if None in [os.environ.get('AWS_ACCESS_KEY_ID'), os.environ.get('AWS_SECRET_ACCESS_KEY')]:
+            errors.append(' - No AWS_ACCESS_KEY_ID or AWS_SECRET_ACCESS_KEY environment vars found')
         else:
-            conn = ec2.connect_to_region(region)
+            errors.append(' - AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment vars found but may not be correct')
 
-        # connect_to_region will fail "silently" by returning None if the region name is wrong or not supported
-        if conn is None:
-            print("region name: %s likely not supported, or AWS is down.  connection to region failed." % region)
-            sys.exit(1)
+        boto_paths = ['/etc/boto.cfg', '~/.boto', '~/.aws/credentials']
+        boto_config_found = list(p for p in boto_paths if os.path.isfile(os.path.expanduser(p)))
+        if len(boto_config_found) > 0:
+            errors.append(" - Boto configs found at '%s', but the credentials contained may not be correct" % ', '.join(boto_config_found))
+        else:
+            errors.append(" - No Boto config found at any expected location '%s'" % ', '.join(boto_paths))
+
+        raise Exception('\n'.join(errors))
```

Comment:
Okay... what about using ``` sys.stderr.write('%s\n' % ('\n'.join(errors))) ``` at the end of this function and then the calling function does sys.exit(1) afterwards?  Almost the same result but it doesn't abuse the exception mechanism to make it happen :-)

Probably should do something similar with the exception raised by the new connect method... maybe make the exception thrown there specific and then in this toplevel exception handler, print the exception message to stderr and sys.exit(1).  The general rule is exceptions should be thrown and unhandled if there are programming errors.  Error messages and sys.exit should be used when there are errors the user can correct without diving into the ansible code.